README.md

```diff
@@ -78,37 +78,47 @@ You might not have to sort the output, if you run through reads
 
 Implement the two algorithms in two tools, `naive` and `lin`, that must be present at the root of the repository once they are built. The test setup checks that they give the correct output on selected data, but you should still carefully test them.
 
-Once you have implemented the tools, fill out the report below. 
+Once you have implemented the tools, fill out the report below.
 
 ## Report
 
-### Insights you may have had while implementing and comparing the algorithms. 
+### Insights you may have had while implementing and comparing the algorithms.
 
-*Describe this here.*
+There are many ways to archive the same thing but not every method is equally elegant.
 
-### Problems encountered if any. 
+### Problems encountered if any.
 
-*Describe this here.*
+* Plotting resonable data can be very hard. Mainly making the plots readable so a conclusion can be made from them is difficult.
+* Indices can also be difficult.
 
 ### Experiments that verifies the correctness of your implementations.
 
-*Describe this here.*
+We have made testing with some examples were we knew the correct answer and checked that the algorithm gave the same answer as we expected.
+
+After being confident in that the naive algorithm as correct we also ran alot of examples with randomly generated data and compared the output given by different algorithms to make sure they all gave the same answer to the same input.
 
 ### Experiments validating the running time.
 
 For this section, you should address the following:
 
-* An experiment that verifies that your implementation of `naive` uses no more time than O(nm) to find all occurrences of a given pattern in a text. Remember to explain your choice of test data. What are “best” and “worst” case inputs? 
-
+* An experiment that verifies that your implementation of `naive` uses no more time than O(nm) to find all occurrences of a given pattern in a text. Remember to explain your choice of test data. What are “best” and “worst” case inputs?
+* We chose to use random data for testing the run-time. Since it is the best approximation of real data. This of course will most likely not give a worst case or best case example to run on. But running a lot of iterations it should all average out. For each seqence length we run 100 seqences of that length with a pattern length of m=100, 1000, 10000 for each of the algorithms. For every unique seqence we take the average time of running the algorithm with the same input 5 times.
```

Comment:
Taking averages helps you pin down an underlying parameter. Here it would be the expected running time for random strings. It doesn't generalise to anything beyond that. It doesn't matter how often you experiment on a random string to get an increasingly more accurate estimate of the expected running time for random strings, it will not tell you what the time will look like on, for example, highly repetitive strings.

The law of large numbers is not magic; averages converge to expectations (and the central limit theorem tells you how). But if you look at the wrong distribution--here, that would be looking at the wrong class of strings--it doesn't tell you anything. That would be wonderful, and life would be a lot simpler, but it sadly isn't so.

For the naive algorithm, random strings are closer to the best case complexity than you might think. If you have an alphabet of size $\sigma$, then the chance of a match when you place `p[0]` somewhere along `x[i]` is $1/\sigma$. (This is not precisely true for the algorithm because characters are not independent in the search, but it is damned close). If the probability of a match is $1/\sigma$, the probability of a mismatch is $1-1/\sigma=\frac{\sigma-1}{\sigma}$, and you have a geometric distribution with waiting time $\frac{\sigma}{\sigma-1}$. That is the expected number of comparisons you will make for each position you place `p` at along `x`.

Notice that we just reasoned that the number of comparisons we do for each position in `x` is constant; we have a running time that says $O(n)$ because $m$ doesn't come into play. What matters is how long we expect to match, now how many characters we have after the first mismatch. The latter is entirely irrelevant.

The math breaks down with a single character alphabet because we are picking random letters and asking when we have a mismatch; it is where the independence also has a problem, but for larger alphabets, we get an idea about the performance from this simple model.

If our alphabet has size 2, we expect to compare $\sigma/(\sigma-1) = 2$ characters per position we place `p` at. That means that for an alphabet with two characters, we expect to run twice as long as the best case scenario where we only compare characters once per position in `x`.

For DNA, $\sigma=4$, we expect to compare $4/3=1.33$ characters. That is, every four time we place `p` along `x`, we will have a mismatch on the first comparison in three of the cases and look at the second character in the fourth.

As the alphabet size grows, the expectation that we look at more than the first character shrinks, and we get closer and closer to the best-case scenario.

Random strings are good approximations of the *best-case* data for this algorithm. Because of that, we would be very unlikely to see worst-case behaviour. For that, we have to work at it. (Or use strings that all consist of the same letter, as you also point out).

Sorry, long rant. I just wanted to make clear that nothing averages out here; it is a dangerous way to think sometimes, and here in particular. In many other cases, it is the right kind of thinking, though.